src/filesystem/implementations/as.h

```diff
@@ -175,72 +177,79 @@ ASFileSystem::CheckClient()
-    std::function<
-        Status(const as::list_blobs_segmented_item&, const std::string&)>
-        func)
+    std::function<Status(
+        const std::vector<asb::Models::BlobItem>& blobs,
+        const std::vector<std::string>& blob_prefixes)>
+        callback)
 {
-  as::blob_client_wrapper bc(client_);
-
+  auto container_client = client_->GetBlobContainerClient(container);
+  auto options = asb::ListBlobsOptions();
   // Append a slash to make it easier to list contents
   std::string full_dir = AppendSlash(dir_path);
-  auto blobs = bc.list_blobs_segmented(container, "/", "", full_dir);
-  if (errno != 0) {
-    return Status(
-        Status::Code::INTERNAL, "Failed to get contents of directory " +
-                                    dir_path + ", errno:" + strerror(errno));
-  }
+  options.Prefix = full_dir;
 
-  for (auto&& item : blobs.blobs) {
-    std::string name = item.name;
-    int item_start = name.find(full_dir) + full_dir.size();
-    int item_end = name.find("/", item_start);
-    // Let set take care of subdirectory contents
-    std::string subfile = name.substr(item_start, item_end - item_start);
-    auto status = func(item, subfile);
-    if (!status.IsOk()) {
-      return status;
+  try {
+    for (auto blobPage = container_client.ListBlobsByHierarchy("/", options);
+         blobPage.HasPage(); blobPage.MoveToNextPage()) {
+      // per-page per-blob
+      RETURN_IF_ERROR(callback(blobPage.Blobs, blobPage.BlobPrefixes));
     }
   }
+  catch (as::StorageException& ex) {
+    return Status(
+        Status::Code::INTERNAL,
+        "Failed to get contents of directory " + dir_path + ":" + ex.what());
+  }
+
   return Status::Success;
 }
 
 Status
 ASFileSystem::GetDirectoryContents(
     const std::string& path, std::set<std::string>* contents)
 {
-  auto func = [&](const as::list_blobs_segmented_item& item,
-                  const std::string& dir) {
-    contents->insert(dir);
-    // Fail-safe check to ensure the item name is not empty
-    if (dir.empty()) {
-      return Status(
-          Status::Code::INTERNAL,
-          "Cannot handle item with empty name at " + path);
+  auto func = [&](const std::vector<asb::Models::BlobItem>& blobs,
+                  const std::vector<std::string>& blob_prefixes) {
+    for (const auto& blob_item : blobs) {
+      // Fail-safe check to ensure the item name is not empty
+      if (blob_item.Name.empty()) {
+        return Status(
+            Status::Code::INTERNAL,
+            "Cannot handle item with empty name at " + path);
+      }
+      contents->insert(BaseName(blob_item.Name));
+    }
+    for (const auto& directory_item : blob_prefixes) {
+      // Fail-safe check to ensure the item name is not empty
+      if (directory_item.empty()) {
+        return Status(
+            Status::Code::INTERNAL,
+            "Cannot handle item with empty name at " + path);
+      }
+      contents->insert(BaseName(directory_item));
```

Comment:
Can you add something to differentiate these two error messages so it's easier to debug which part they originated from? blob_item vs directory_item